verilog/tools/ls/autoexpand.cc

```diff
@@ -0,0 +1,289 @@
+                            const std::vector<absl::string_view> &port_names,
+                            std::ostream &output);
+
+  // Find the location and span of an AUTO comment given
+  absl::optional<AutoComment> FindInSpan(const std::regex &re,
+                                         absl::string_view span);
+
+  // Retrieves port names from a module declared before the given offset
+  std::set<absl::string_view> GetPortsDeclaredBefore(const Symbol &module,
+                                                     int offset);
+
+  // Expand AUTOARG for the given module (skip given port set)
+  std::string ExpandAutoarg(
+      const Symbol &module,
+      const std::set<absl::string_view> &predeclared_ports);
+
+  // Expand all AUTOs in the buffer
+  std::vector<TextEdit> Expand();
+
+ private:
+  // Text structure of the buffer to expand AUTOs in
+  const TextStructureView &text_structure_;
+
+  // Token context created from the buffer contents
+  const TokenInfo::Context context_;
+
+  // String to add at the end of each generated line
+  std::string line_end_str_;
+
+  // Regex for finding AUTOARG comments
+  static std::regex autoarg_re_;
+};
+
+std::regex AutoExpander::autoarg_re_{R"(/\*\s*AUTOARG\s*\*/)"};
+
+void AutoExpander::EmitPortDeclarations(
+    const absl::string_view header,
+    const std::vector<absl::string_view> &port_names, std::ostream &output) {
+  if (port_names.empty()) return;
+  output << line_end_str_ << "// " << header << line_end_str_;
+  bool first = true;
+  for (auto port : port_names) {
+    if (!first) {
+      output << ',' << line_end_str_;
+    }
+    first = false;
+    output << port;
+  }
+}
+
+absl::optional<AutoExpander::AutoComment> AutoExpander::FindInSpan(
+    const std::regex &re, const absl::string_view span) {
+  std::match_results<const char *> match;
+  if (!std::regex_search(span.begin(), span.end(), match, re)) return {};
+  const int start_offset =
+      std::distance(context_.base.begin(), span.begin() + match.position());
+  return {{.linecol = text_structure_.GetLineColAtOffset(start_offset),
+           .start_offset = start_offset,
+           .contents = {span.begin() + match.position(),
+                        static_cast<size_t>(match.length())}}};
+}
+
+std::set<absl::string_view> AutoExpander::GetPortsDeclaredBefore(
+    const Symbol &module, const int offset) {
+  std::set<absl::string_view> ports_before;
+  const auto all_ports = GetModulePortDeclarationList(module);
+  if (!all_ports) return {};
+  for (const SymbolPtr &port : all_ports->children()) {
+    if (port->Kind() == SymbolKind::kLeaf) continue;
+    // Find the identifier of the port
+    const SyntaxTreeNode &port_node = SymbolCastToNode(*port);
+    const NodeEnum tag = NodeEnum(port_node.Tag().tag);
+    const SyntaxTreeLeaf *port_id_node = nullptr;
+    if (tag == NodeEnum::kPortDeclaration) {
+      port_id_node = GetIdentifierFromPortDeclaration(port_node);
+    } else if (tag == NodeEnum::kPort) {
```

Comment:
Mutant testing find:
Changing this 1 line to
    } else if (true) {
does not cause any test exercising them to fail.